### PR TITLE
Implement configurable JSON encoder/decoder to allow non-serializable python-action return values

### PR DESCRIPTION
**Pull request by rbdixon**
Per #98.

- [x] Add encoder/decoder parameters to JSON backend
- [x] Add encoder/decoder parameters to gdbm backend
- [x] Add encoder/decoder parameters to sqlite backend
- [x] Add encoder/decoder parameter to `Dependency` manager
- [x] Add tests for backends
- [x] Add `get_encoder_decoder()` function to `DoitCmdBase` to use config parameters to specify JSON encoder/decoder
- [x] Pass encoder/decoder value to `Dependency` manager instance in `cmd_base.py`
- [x] Document new configuration parameter

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -76,27 +76,31 @@ def remove_db(filename):
...
+    name = request._pyfuncitem.name
+    name = py.std.re.sub("[\W]", "_", name)
+    return request.config._tmpdirhandler.mktemp(name, numbered=True)
+
+@pytest.fixture
+def dep_manager(request, dep_tmpdir):
+    # TODO: This doesn't work as expected
+    # Migrate to test_dependency.py fixture
     if hasattr(request, 'param'):
+        raise Exception('Did not expect this code to be used... investigate')
```
Comment (rbdixon) [inline tests/conftest.py]: @schettino72 I _think_ this fixture was attempting to inspect the `request.param` attribute which is in use on the fixture parametrization on `test/test_dependency.py#69` and then setup the requested backend. pytest does not currently work like that. Maybe it used to?

To verify I put that exception in just to see if anything hit it. Nothing ever does.

I duplicated this fixture in `tests/test_dependency.py` and moved the backend setup code there.

I think that in prior versions only the default backend was being tested in `tests/test_dependency.py`.

**Comment by rbdixon** [comment]
This is usable and has working tests. I'm going to try it out here for a while and see how it goes.

File: tests/conftest.py (hunk 1 above)
Comment (schettino72) [reply tests/conftest.py]: good catch. thanks. fixed on master.

**Comment by rbdixon** [comment]
Rebased. I'll squash this whole series of patches into one commit once it is ready.

**Comment by schettino72** [comment]
thanks,  usually i squash myself when merging.

if you squash you mess up with inline comments here. and if requires another review you would need to sqaush again..  not productive...

**Comment by rbdixon** [comment]
I think this is ready to go. I’ve been using it locally for a few days. 

I’m offline for a week so no rush merging it. If you would like me to do more on this let me know. Thank you!

**Comment by schettino72** [comment]
It is not good practice to reformat the code... makes code-review harder. But I am guilty of having some really bad formatting there (apart from my decision to ignore PEP8 regarding vertical space), so I will fix the mess myself.

I thought it was interesting that you wanted to use pickle...  putting pickle inside JSON is ugly as hell, but why not support pickle as first class? I want to explore this and might spin this PR in a different direction.

Anyway I will handle it myself, enjoy your time off :smile:

**Comment by rbdixon** [comment]
Oh... I’m sorry. I use a formatter on all my work projects and sometimes it creeps into non work projects. I can back that all out if you would prefer. 

If you ever want to consider a formatter black is very nice.

**Comment by rbdixon** [comment]
I’m back from vacation. Anything I can help with to get this merged?

**Comment by schettino72** [comment]
can you please rebase to master and remove the formatting changes?

on master i fixed some formatting problems.... i am not a fan.of 'black' for formatting as i use blank lines different from pep8 recommendation.

**Comment by rbdixon** [comment]
OK, I think I've got all of the changes black made to the test directory reverted.

All tests pass _except_ these tests which also fail on master:

- `tests/test_cmd_auto.py:TestAuto.test_run_wait`
- `tests/test_filewatch.py:TestFileWatcher.testLoop`: This test seems to get stuck waiting.

I hope this is helpful. thank you!

**Comment by rbdixon** [comment]
How did the `black` reversions look?

**Comment by schettino72** [comment]
Sorry I had a review of this that was not published. I have merged the functionality, my changes:

- need to review docs ( I ran out of time, but merged anyway)
- I combined encoder/decoder in a single "codec" parameter (no need to be JSON based!)
- I removed support to pass a class as a string. I have already written code to support this for plugins. So this should be a plugin...
- Also possible to pass a class directly on DOIT_CONFIG, example below. 
- I removed JSONNullEncoder
- I dont think test_dependency really needs to test codecs...
- fixed strace tests that were broken


```
# encoder for db backend
class JSONNullEncoder(json.JSONEncoder):
    'Encodes non-serializable values as null'

    def default(self, obj):
        if type(obj) not in [str, int, float, bool, list, dict, None]:
            return None
        else:
            return obj

class MyCodec(JSONCodec):
    def __init__(self):
        self.encoder = JSONNullEncoder()
        self.decoder = json.JSONDecoder()

DOIT_CONFIG = {
    'codec_cls': MyCodec,
}
```

**Comment by schettino72** [comment]
leaving this open until I merge docs.

**Comment by rbdixon** [comment]
I gave this a try and it works as expected. Thank you.

The only problem I see is a minor typo in `CHANGES`: `BACKWARS INCOMPATIBLE`

This is really helpful for my application of doit.

**Pull request closed without merging**